Feature/494 new group model

quiteqiang opened the pull request:
Group Module API and tests --Done

PROgram52bc committed:
moved Manager from people to groups (flask command working)

PROgram52bc committed:
modified groups.models to match the new ERD

PROgram52bc committed:
added factory functions for group_type and manager_type

PROgram52bc committed:
implemented CRUD api and test for group_types

PROgram52bc committed:
removed dependency on create_multiple_managers in faker.py

PROgram52bc committed:
implemented api and test for ManagerType

PROgram52bc committed:
added test code (commented out) for asset that will fail

PROgram52bc committed:
removed failing tests from test_groups.py

PROgram52bc committed:
added create_group endpoint

PROgram52bc committed:
fixed create_group to return a json object

PROgram52bc committed:
added repr methods to GroupType and ManagerType

PROgram52bc committed:
added read_all_groups without query filtering

PROgram52bc committed:
added read_one_group endpoint

PROgram52bc committed:
removed find_group endpoint

Commit:
test_create_group/read_all_groups/read_one_group

Commit:
Merge branch 'feature/494-new-group-model' of https://github.com/corpus-christi/corpus-christi into feature/494-new-group-model-test

PROgram52bc committed:
added update_group endpoint

Commit:
Merge branch 'feature/494-new-group-model' of https://github.com/corpus-christi/corpus-christi into feature/494-new-group-model-test
"update change"

PROgram52bc committed:
fixed update_group

Commit:
Merge branch 'feature/494-new-group-model' of https://github.com/corpus-christi/corpus-christi into feature/494-new-group-model-test
update

PROgram52bc committed:
implemented query string filtering functionality (except for 'order')

Commit:
Group table tests done

Commit:
Merge branch 'feature/494-new-group-model-test' into feature/494-new-group-model

Commit:
create_group_data update

PROgram52bc committed:
added create_multiple_groups to flask app groups command

PROgram52bc committed:
implemented 'order' query, get_all_queried_entities using custom error
object

PROgram52bc committed:
read_all_group_types and read_all_manager_types support query

PROgram52bc committed:
create_manager implemented

PROgram52bc committed:
added create_multiple_managers

Commit:
tests for Retrieveing

PROgram52bc committed:
added create_manager and read_all_managers

PROgram52bc committed:
Merge branch 'feature/494-new-group-model' of github.com:corpus-christi/corpus-christi into feature/494-new-group-model

PROgram52bc committed:
removed non-existing dependency in test_groups.py

PROgram52bc committed:
updated models, removing ids from managers and members

Commit:
create & add more managers

PROgram52bc committed:
added read_one_manager, update_manager, delete_manager

PROgram52bc committed:
fixed create_multiple_managers repeating primary key bug

PROgram52bc committed:
changed create_multiple_managers to take a fraction

PROgram52bc committed:
fixed fraction error in create_multiple_managers

PROgram52bc committed:
removed 'id' reference in GroupSchema

Commit:
test read one manager

PROgram52bc committed:
use exclude fields in update_manager

PROgram52bc committed:
added type validation on route parameter

PROgram52bc committed:
updated groups/api.py

- implemented CRUD pattern for Member table
- modified delete endpoints to return 204 with an empty body (a string)
- return 409 error upon creating an existing entity
- more strict payload validation (e.g. group_id field invalid on create_manager)

PROgram52bc committed:
updated create_group_data.py

- added create_multiple_members
- added member_object_factory

PROgram52bc committed:
updated groups/models

added nested fields in Meeting schema

PROgram52bc committed:
updated groups/api

- 'active' is now an optional field default to True in create_* routes
- added delete_group route
- implemented CRUD endpoints for Meeting

PROgram52bc committed:
added meeting_object_factory and create_multiple_meetings

PROgram52bc committed:
create_multiple_attendance working

PROgram52bc committed:
reduced country-code size to speed up testing

'mv api/src/places/data/country-codes-full.json api/src/places/data/country-codes.json'
    any time to restore to the original country code data

PROgram52bc committed:
updated groups/api

- CRUD for attendances

Commit:
all the tests done

Commit:
pass all tests

PROgram52bc committed:
Merge branch 'development' into feature/494-new-group-model

Commit:
fixed delete_attendance

PROgram52bc committed:
Merge branch 'feature/494-new-group-model' of github.com:corpus-christi/corpus-christi into feature/494-new-group-model

nurkkala commented on api/commands/faker.py:
```diff
@@ -45,8 +45,8 @@ def fake_events():
-        create_multiple_managers(db.session, 2, 'Group Overseer')
-        create_multiple_managers(db.session, 5, 'Group Leader')
+        # create_multiple_managers(db.session, 2, 'Group Overseer')
+        # create_multiple_managers(db.session, 5, 'Group Leader')
```
If we intend to uncomment these lines and use them again, please include a comment to that effect.

If we don't intend to use these lines again, go ahead and just delete them.

PROgram52bc replied:
will be deleted in new commits!

nurkkala commented on api/src/assets/test_assets.py:
```diff
@@ -42,6 +42,19 @@ def test_create_asset(auth_client):
     for attr in new_asset:
         assert new_asset[attr] == queried_asset.__dict__[attr]
 
+# # Currently will fail, since the api does not check for invalid location
```
Rather than comment out a test, use the `skip` decorator from pytest.
See https://docs.pytest.org/en/latest/skipping.html

PROgram52bc replied:
Will be fixed!

nurkkala commented on api/src/auth/api.py:
```diff
@@ -58,6 +58,7 @@ def login():
 # Define our callback function to check if a token has been revoked or not
 @jwt.token_in_blacklist_loader
 def check_if_token_revoked(decoded_token):
+    return False # for testing purpose
```
It doesn't seem like this line should be left in the code.

If we need to override this method for testing, can we use a mock? For example, https://docs.pytest.org/en/latest/monkeypatch.html

PROgram52bc replied:
Sorry about that. I will double check next time to make sure code like this does not get committed. I was just doing some quick-and-dirty testing using an interactive API tester, and somehow forgot to remove this line... I will remove it in a new commit.

nurkkala replied:
No worries -- it's easy to miss these small changes.

nurkkala commented on api/src/courses/test_courses.py:
```diff
@@ -688,7 +688,7 @@ def test_create_course_offering(auth_client):
 def test_read_all_course_offerings(auth_client):
     # GIVEN existing (active and inactive) course offerings
-    create_multiple_courses(auth_client.sqla, 1)
+    create_multiple_(auth_client.sqla, 1)
```
This looks like an unintentional change.

PROgram52bc replied:
will be fixed

nurkkala commented on api/src/groups/api.py:
```diff
@@ -5,527 +5,507 @@
+def read_one_group_type(group_type_id):
+    group_type = db.session.query(GroupType).filter_by(id=group_type_id).first()
+    if not group_type:
+        return jsonify(f"GroupType with id #{group_type_id} does not exist."), 404
```
This construct of `jsonify(...some text ...)` appears fairly often in the code.  It seems kind of weird, in that a string is _already_ JSON. I wonder if this pattern has seeped into the code from previous developers using copy-and-paste? Take a look at how the UI deals with this kind of response. Is there a way to make it more general? It might be best to create an new issue so we remember to inspect this throughout the code base.

PROgram52bc replied:
Yeah. I believe this kind of construct has been in the code base for a pretty long time. I took that construct from the `events` module. And after a quick glance at the UI code, I believe in many places, the UI simply leaves error messages unhandled, so it won't make much of a difference what kind of object we return from the API upon error. But anyhow, do you have any recommendation regarding how we should change that construct? Maybe just return a string instead?

nurkkala replied:
This is going to require more thought.

nurkkala commented on api/src/groups/api.py:
```diff
@@ -5,527 +5,507 @@
 
-    today = datetime.datetime.today().strftime('%Y-%m-%d')
+    if not group_type:
+        return jsonify(f"GroupType with id #{group_type_id} does not exist."), 404
```
Here's another instance of that `jsonify(... text ...)` thing.

PROgram52bc replied:
yes..

nurkkala commented on api/src/groups/api.py:
```diff
@@ -5,527 +5,507 @@
+    return jsonify(group_schema.dump(new_group)), 201
 
 
 @groups.route('/groups')
```
Add `methods=['GET']`, here and throughout where appropriate.

PROgram52bc replied:
will be fixed

nurkkala commented on api/src/groups/api.py:
```diff
@@ -5,527 +5,507 @@
-    if group is None:
-        return jsonify(msg="Group not found"), 404
-
-    new_manager_id = None
```
Is this code going away entirely? If so, is that the case because of the database changes?

PROgram52bc replied:
I think the `modify_entity` function will return a 404 error in case the group specified does not exist. Not sure if that's the best practice, but I noticed that this `modify_entity` function is used throughout the `event` module and thought it's probably better to reuse something that's already there... We should definitely change the interface to be more transparent, if needed.

nurkkala replied:
Agreed.

nurkkala commented on api/src/groups/api.py:
```diff
@@ -5,527 +5,507 @@
+    if not manager_type:
+        return jsonify(f"ManagerType with id #{manager_type_id} does not exist."), 404
 
+    db.session.delete(manager_type)
```
In order to delete a manager type, we would have to make sure no `Manager` records were using that type. Should we:
* Only allow this if there are no managers using this type?
* Make the manager type FK be nullable in the `Manager` model class?
* Allow the user to choose a different manager type that will replace instances of the manager type that we're about to delete?

PROgram52bc replied:
Yeah, I think you made a good point... Currently it will probably fail with an internal server error (and the samething for a `GroupType`). 

For the three options, here are just some of my thoughts:
1. I can see this mechanism working, but it seems a little strict, in that when someone tries to modify `manager_types` they would have to remember to first make a new type and change all managers to that type _before_ deleting the existing type. But I guess we can do something in the UI to make it more user-friendly as well.
1. This is just an alternative to the previous option, which offers more flexibility.
1. This can perhaps be a logic on the UI side, where the user is prompted to enter a new name for the manager type to be deleted. On the API side, this can be done by making a PATCH request to the same endpoint.

So it seems to me that for API, the decision is just between the first and second, about whether we need some extra flexbility by making the FK nullable. What do you prefer?
I will implement the change once we reach a conclusion on the design.

nurkkala replied:
Good analysis. How about:
1. Server side: leave things alone; it's good to have the referential integrity that you get with a non-nullable FK.
1. UI side: if there are instances of `Manager` with the given `Manager Type`, implement UI that allows the user to change those instances to be of a different `Manager Type`.

The UI could, for example, list everyone who has the "doomed" type, along with a dropdown of other management types, allowing the user to choose a new management type for each such manager. A simpler alternative would be to only allow a "bulk" change, where _everyone_ with the doomed type would be updated to a different type.

Which UI approach do you think would be better for the end user?

We need a new issue for this change.

nurkkala commented on api/src/groups/api.py:
```diff
@@ -5,527 +5,507 @@
 
 
-@groups.route('/meetings/address/<address_id>')
+@groups.route('/groups/<int:group_id>/managers', methods=['POST'])
```
I like this approach to the URL.

nurkkala commented on api/src/groups/api.py:
```diff
@@ -5,527 +5,507 @@
-    return jsonify(meeting_schema.dump(result, many=True))
+    if db.session.query(Manager).filter_by(person_id=person_id, group_id=group_id).first():
+        # if the same manager already exists
+        return jsonify(f"Manager with group_id #{group_id} and person_id #{person_id} already exists"), 409
```
This error message is not going to be very helpful for the end user. Rather than talking about database IDs that have no meaning to the user, could this be changed to say something like "Fred Ziffle is already a manager of the Young Families group."

PROgram52bc replied:
Yeah. That makes total sense. And like the `jsonify(...some text ...)` construct, I think this is also a common practice in the code base since a long time ago. Do we want to combine those problems into an issue and make change to all of them at once?

nurkkala replied:
Yes, let's create an issue (if you've not done so already). This PR is not the right place to resolve the question.

nurkkala commented on api/src/groups/api.py:
```diff
@@ -5,527 +5,507 @@
+def read_one_manager(group_id, person_id):
+    manager = db.session.query(Manager).filter_by(group_id=group_id, person_id=person_id).first()
+    if manager is None:
+        return jsonify(f"Manager with group_id #{group_id} and person_id #{person_id} does not exist"), 404
```
Similarly confusing error message. Please create an issue that asks for all error messages in CC to be reviewed to try to make them as helpful to the end user as possible.

PROgram52bc replied:
Will do that. But just a thought: if the the APIs are primarily to be consumed by the UI, should the UI be responsible for making a request to a valid entity? Do we intend to display the error from the API directly in the user interface rather than transforming it with some extra logic? If not, I think error messages like this can be more helpful in debugging, even though it does not make any sense for an end user.

nurkkala replied:
We want both useful UI and info for debugging. I think this is covered by the plan to improve error reporting to the end user and adding a logging mechanism on the server side.

nurkkala commented on api/src/groups/api.py:
```diff
@@ -5,527 +5,507 @@
-    for member in meeting.members:
-        db.session.delete(member)
-    db.session.delete(meeting)
+    db.session.delete(manager)
```
Although this is a "delete" operation, we very seldom want to delete data from the database. Instead, set the `active` field to `False`. Please create an issue to review delete endpoints the entire code base; there _may_ be some cases in which we really do want to delete, but let's be sure that we are using `active` flags anywhere that we don't want to lose data forever.

PROgram52bc replied:
Okay, I see what you are saying. So do we want to implement all the DELETE endpoints with an `inactivate` operation? If so, should we change all the `read_all` endpoints to return only entities that are active?
I'm thinking if we keep the current endpoints, we can still make a PATCH request with {'active': 'false'} in the UI when we want to "archive" some entity, and only make a DELETE request when we _really_ want to delete it.

nurkkala replied:
The default design should be to use an `active` flag; there may be some cases where we want to actually delete something, but I can't think of any right away. Maybe something like whether a group member has attended a meeting could actually be deleted, for example, if the person taking attendance just made a mistake. But this should be the exception.

I like your idea of using `PATCH` instead of `DELETE`. I've thought of `DELETE` operations with an `active` flag as kind of a "virtual delete," but the use of `PATCH` here does seem like it's more in the spirit of RESTful API design.

nurkkala commented on api/src/groups/api.py:
```diff
@@ -5,527 +5,507 @@
+        return jsonify(f"Meeting with id #{meeting_id} does not exist"), 404
 
-    setattr(meeting, 'active', False)
+    db.session.delete(meeting)
```
Here's a case where in some cases, we may actually want to delete a meeting. For example, if a meeting was scheduled and never held and nobody was marked as attending the meeting, it was _probably_ created in error and could be removed. However, if the meeting took place, had attendees, etc., it should probably not be allowed to be deleted.

PROgram52bc replied:
Like I said in the previous comment, I'm _personally_ still leaning toward making the API endpoints as explicit as possible (e.g. a DELETE endpoint actually _deletes_ the entity), and implement some "trash can" functionality in the UI, utilizing those endpoints. 
But we can totally change it to the other way, if needed!

nurkkala replied:
Let's go with your plan to be explicit about the endpoints.

Please update the RESTful documentation in the `coding.md` document to reflect this change (i.e., use `PATCH` unless you _really_ want to something to go away, in which case use `DELETE`).

nurkkala commented on api/src/groups/api.py:
```diff
@@ -5,527 +5,507 @@
+    try:
+        db.session.commit()
+    except IntegrityError:
+        return jsonify('Payload contains an invalid foreign key'), 404
```
This message appears numerous times in this one file. As mentioned, if we are gong to report such errors, they should have some meaning to the end user. For example, we could figure out which FK was invalid and try to map that back into a message that would be helpful.

Another problem with these error messages is that they are not currently internationalized. Please create an issue to review these messages and come up with I18N information for them. In particular, we'll have to figure out how to include parameters in the messages rather than just fixed text in the localization.

nurkkala replied:
As we discussed in the stand-up, please open a new issue for error handling and reporting (if you've not done so already).

PROgram52bc replied:
see #530

nurkkala commented on api/src/groups/api.py:
```diff
@@ -5,527 +5,507 @@
-        return jsonify(err.messages), 422
+def create_attendance(meeting_id, person_id):
+    if not db.session.query(Person).filter_by(id=person_id).first():
+        return jsonify(f"Person with person_id #{person_id} does not exist"), 404
```
Thinking about these error messages, we don't want to report internal database keys to the end user, but we _do_ probably want to log these kinds of errors someplace. I'd bet that there are several standard packages for logging errors and warnings to a file that could be inspected by the system administrator. I've created an issue about this.

nurkkala commented on api/src/groups/test_groups.py:
```diff
@@ -5,993 +5,816 @@
+    # THEN assert exception thrown
     assert resp.status_code == 422
-
+    # Test creating valid group
```
In general, tests should test _just one thing_. Split this test into two: `test_create_invalid_group` and `test_create_valid_group`, for example.

quiteqiang replied:
Got it! will fix it soon

nurkkala commented on api/src/groups/test_groups.py:
```diff
@@ -5,993 +5,816 @@
+        url_for('groups.update_group', group_id=555))
+    # THEN assert error code
+    assert resp.status_code == 422
+    # Test with populated database
```
Refactor into its own test function.

nurkkala commented on api/src/groups/test_groups.py:
```diff
@@ -5,993 +5,816 @@
+    create_multiple_group_types(auth_client.sqla, 1)
+    count = random.randint(3, 11)
+    create_multiple_groups(auth_client.sqla, count)
+    # Test with populated database
```
Refactor into own test function. 

Take a look for other places where multiple tests have been lumped into a single function and separate them.

nurkkala commented on api/src/groups/test_groups.py:
```diff
@@ -5,993 +5,816 @@
+#     pass
+#
+@pytest.mark.smoke
+def test_subset_group(auth_client):
```
This test function is way too long. Separate into individual tests. If there is common test setup, split that into a separate function.

quiteqiang replied:
Will fix it soon

nurkkala commented on api/src/groups/test_groups.py:
```diff
@@ -5,993 +5,816 @@
-        assert auth_client.sqla.query(Meeting).filter_by(**filtered_meeting).count() == 1
-
-
-def generate_addresses(auth_client, count=1):
```
Why has this been replaced with a `pass` function?

Are these tests no longer required?

PROgram52bc replied:
I think the `generate_addresses` function can be replaced by the `create_multiple_addresses` function from the places module, and that's why we don't need a separate function to create addresses.

nurkkala commented on api/src/groups/test_groups.py:
```diff
@@ -5,993 +5,816 @@
-    assert resp.status_code == 404
-
-
-def test_activate_meeting(auth_client):
```
Why are these tests deleted? Did they move somewhere else? Are they no longer relevant?

PROgram52bc replied:
Since there are many changes made to the database related to the `Group` module, I figured that most of those tests would need to be rewritten as well, and therefore removed all of them at once. For example, `test_read_all_meetings_by_group` is no longer relevant since we no longer have an endpoint called `read_all_meetings_by_group` now. But there are perhaps parts that are better modified than removed and rewritten, which was my fault...
The `read_all_meetings_by_group` functionality can now be achieved with adding a query string to the "read_all" endpoint. To test that the query strings work properly (e.g. `...?where=group_id:4`), @quiteqiang implemented some of the tests in `test_subset_group`; but since we now have a general mechanism to parse the query string, do you think we should write some general tests that test on the query string functionality, or should we rewrite them for all entities?
@quiteqiang is working on splitting large test functions into smaller ones, and adding tests for other status codes, as well as adding some of the tests that are still relevant now (like `test_activate_meeting`).

nurkkala replied:
Regarding the tests for parsing queries, we should have comprehensive tests for that general functionality, but we _don't_ need to repeat those tests everywhere they're used.

nurkkala commented on api/src/groups/test_groups.py:
```diff
@@ -5,993 +5,816 @@
     assert resp.status_code == 404
 
+# 
+# @pytest.mark.smoke
```
Tests with this decorator are called "smoke tests." A smoke test is supposed to be a test that will run quickly to give the developer rapid feedback that the system is basically working. The idea is that you want to be able to run these tests frequently; if they take too long, the developer will tend _not_ to run them. Less often, you'll run _all_ the tests to get comprehensive test results.

So don't necessarily mark tons of tests as smoke tests. We should have a few smoke tests in for each module, but not 10's or 100's of them.

quiteqiang replied:
Sorry about this, I thought this is just a simple marker for tester code. Will fix this issue soon

nurkkala replied:
No worries. Note that this "tagging" method is completely flexible. If we wanted to create other subsets of the tests, we could add new tags that would let us test different subsystems more easily. I haven't gone in this direction yet, because you can also run tests by file name, etc.

nurkkala commented on api/src/groups/test_groups.py:
```diff
@@ -5,993 +5,816 @@
+# def test_repr_group(auth_client):
+#     pass
+# 
+# @pytest.mark.smoke
```
This group of unimplemented tests should be marked with the `skip` decorator so that the fact that they're not being run gets reported when we run all the tests. That way we won't forget to go back and implement the tests.

nurkkala commented on api/src/groups/test_groups.py:
```diff
@@ -5,993 +5,816 @@
+    create_multiple_groups(auth_client.sqla,1)
+    create_multiple_manager_types(auth_client.sqla,1)
+    create_multiple_managers(auth_client.sqla, 0.75)
+    # WHEN we get the first manager
```
The `GIVEN`, `WHEN`, `THEN` pattern is supposed to read like a normal sentence. Normally when there are multiple `WHEN`s that add to one another, I write `WHEN` for the  first one then `AND` for all subsequent. So it would end up `GIVEN an empty database`, `WHEN we add managers`, `AND we get the first manager`, ... `THEN ...`.

nurkkala commented on api/src/groups/test_groups.py:
```diff
@@ -5,993 +5,816 @@
+#
+# @pytest.mark.smoke
+# def test_delete_manager_with_subordinate(auth_client):
+#     pass
```
As mentioned, mark these as skipped.

nurkkala commented on api/src/places/data/country-codes-full.json:
```diff
@@ -0,0 +1,2652 @@
+      },
+      {
+        "locale_code": "es-EC",
+        "name": "Afganist\u00c3\u00a1n"
```
Did something strange happened here? What are these Unicode characters??

PROgram52bc replied:
I only did a rename from `country-codes.json` to `country-codes-full.json` so that the related test will take less time to complete. I believe those unicode characters are the accented characters or characters in other languages.

nurkkala commented on api/src/shared/helpers.py:
```diff
@@ -28,6 +32,89 @@ def get_exclusion_list(query_object, default_exclusion_list):
             ret_list.remove(exclusion)
     return ret_list
 
```
NICE!

nurkkala reviewed (commented):
This looks great!

I've made some comments that I'd like you to address. Most of the comments ask questions or request changes. Don't in any way interpret the comments as a negative -- I just want to make your already good work even better!

PROgram52bc committed:
fixed review comments

- removed unnecessary comments
- marked test_create_asset_with_invalid_location as skip
- re-enabled black list
- fixed typo in test_courses.py
- added explicit GET in endpoints

Commit:
Fixed review comments

- Delete redundant code
- Seperate test functionalities into different functions
- Delete extra test smokes
- Add tests to handel more error types

PROgram52bc committed:
updated GroupImage api and tests

- restored test files for the endpoints
- added 'groups' prefix to the endpoints
- added 'images' attribute to GroupSchema
- added read_all_images endpoint
- fixed create_image_data

nurkkala reviewed (commented):
The updated code looks great. Tests are much cleaner; well done!
